Std/Tactic/Basic.lean

```diff
@@ -162,3 +162,27 @@ example (P : (Nat → Nat) → Prop) : P (fun n => n - n) := by
 -/
 macro (name := Conv.equals) "equals " t:term " => " tac:tacticSeq : conv =>
   `(conv| tactic => show (_ = $t); next => $tac)
+
+/-- `lemma` is not supported, please use `theorem` or import `Mathlib.Tactic.Basic` -/
```

Comment:
Why should something in Std instruct to use Mathlib?

Reply:
The idea here is:
* Some users will type `lemma` from muscle memory, and the first half of the warning tells them that they should be writing `theorem` instead
* Some of these users might have just forgotten to import mathlib (where `lemma` is encouraged), and pointing out that this is allowed in mathlib is helpful to them.

Of course it would all be simpler if mathlib didn't have `lemma` either, but that ship has long since sailed.

Reply:
> Some of these users might have just forgotten to import mathlib (where lemma is encouraged), and pointing out that this is allowed in mathlib is helpful to them.

For the "forgot to import mathlib" case, it would be more helpful to have this in Init rather than Std.

Reply:
I think the confusion for me is that it makes it sound like Std is against using lemma. If Mathlib.Init doesn't work then I would rather have the plain implementation of lemma in Std for people to use, and simply have a style rule not to use it internally.

Reply:
Std *is* against using `lemma`, or more precisely it does not want the syntax to diverge from core and if you want lemma everywhere you should take it upstream. Non-implementing `lemma`  here is a way to improve the error message while still not changing the official grammar or causing lean code in the wild to accrue uses of `lemma`. 

`rustc` does this kind of thing a lot, for example if you write function implementations inside a `struct` it will parse the syntax even though it is not legal, specifically so it can tell you that unlike languages like C# and C++, function implementations go in a separate impl block. [(demo)](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=b5dfe53c49452c52936cb1ee682442b2)

Reply:
There's a miscommunication here. I was trying to say that: (1) Std doesn't use `lemma` and (b) Std doesn't have an opinion whether or not `lemma` should be used as a synonym for `theorem`.

My objection is that: (1) this PR makes it seem that (b) is not true and (2) it makes it harder for users to implement `lemma` in their own project.

In my opinion, the best solution is for Std to have no handling for `lemma` at all (satus quo). However, this is a "paper cut" and, as such, the next best idea is probably to implement `lemma` as a convenient synonym for `theorem` for downstream users but _not_ use it internally.

Reply:
> (2) it makes it harder for users to implement `lemma` in their own project.

How do you figure? AFAIK if the user has an implementation of `lemma` like mathlib's then it will continue to work in the same way. The std non-implementation does not prevent downstream implementations. It even makes them slightly easier since it's now just `elabLemma' true`